src/Microsoft.AspNetCore.Routing.Abstractions/RouteData.cs

```diff
@@ -84,5 +85,99 @@ public RouteValueDictionary Values
                 return _values;
             }
         }
+
+        public RouteDataSnapshot PushState(IRouter router)
+        {
+            if (router == null)
+            {
+                throw new ArgumentNullException(nameof(router));
+            }
+
+            return PushState(router, values: null, dataTokens: null);
+        }
+
+        public RouteDataSnapshot PushState(IRouter router, RouteValueDictionary values, RouteValueDictionary dataTokens)
+        {
+            if (router == null)
+            {
+                throw new ArgumentNullException(nameof(router));
+            }
+
+            var snapshot = new RouteDataSnapshot(this);
+            _routers.Add(router);
+
+            if ((_values == null || _values.Count == 0) && values != null)
+            {
+                _values = new RouteValueDictionary(values);
+            }
+            else if (values == null)
+            {
+                // Do nothing
+            }
+            else if (_values != null && _values.Count > 0)
+            {
+                _values = new RouteValueDictionary(_values);
+
+                if (values != null && values.Count > 0)
+                {
+                    foreach (var kvp in values)
+                    {
+                        if (kvp.Value != null)
+                        {
+                            _values[kvp.Key] = kvp.Value;
+                        }
+                    }
+                }
+            }
+
+            if ((_dataTokens == null || _dataTokens.Count == 0) && dataTokens != null)
+            {
+                _dataTokens = new RouteValueDictionary(dataTokens);
+            }
+            else if (dataTokens == null)
+            {
+                // Do nothing
+            }
+            else if (_dataTokens != null && _dataTokens.Count > 0)
+            {
+                _dataTokens = new RouteValueDictionary(_dataTokens);
+
+                if (dataTokens != null && dataTokens.Count > 0)
+                {
+                    foreach (var kvp in dataTokens)
+                    {
+                        _dataTokens[kvp.Key] = kvp.Value;
+                    }
+                }
+            }
+
+            return snapshot;
+        }
+
+        public struct RouteDataSnapshot
+        {
+            private readonly RouteValueDictionary _dataTokens;
+            private readonly int _routerCount;
+            private readonly RouteValueDictionary _values;
+
+            public RouteDataSnapshot(RouteData routeData)
+            {
+                _dataTokens = routeData._dataTokens;
+                _routerCount = routeData.Routers.Count;
+                _values = routeData._values;
+            }
+
+            public void Restore(RouteData routeData)
+            {
+                routeData._dataTokens = _dataTokens;
+                routeData._values = _values;
+
+                var routersToTrim = routeData._routers.Count - _routerCount;
```

Comment:
Is this just being very defensive? In a normal case wouldn't it always just need to remove exactly 1 router?